### DISCUSSION TITLE
Bonsai-SLEAP Recording Visualizers with Tracking Points as Video

### DISCUSSION THREAD

**Discussion by frr665** [bonsai-rx/bonsai#1404] [Q&A] [answered]
Hello Bonsai Users,

I'm currently working on a project that involves marker-less pose estimation using the Bonsai-SLEAP package. Specifically, I am using the PredictSinglePose operator to track various body parts of flies in real-time.

As part of my project, I would like to record a video that includes the tracking dots superimposed on the fly body parts, similar to what we can see bellow with the PoseVisualizer. However, I have encountered a couple of issues in achieving this:

1. When I place a videowriter operator immediately after the PredictSinglePose operator, I receive an error message stating "no method overload found for the given arguments."
2. When I try to externalize the Image output, the resulting video does not contain the tracking points.

I am wondering if there is a way to directly record Bonsai visualizers as .avi files while preserving the tracking points. I would greatly appreciate any insights or suggestions. 

Thanks for your time and assistance!

Best regards,
Hugo

![image](https://github.com/bonsai-rx/bonsai/assets/117995542/f78fd26e-8e5d-44c5-84aa-66610cb94e57)

**Comment by bruno-f-cruz** [comment, marked as answer]
If you just want to demo it, the easiest way would be to use `ScreenCaptureStream`. Otherwise, you can assemble the full visualizer yourself using each one of the points using [OpenCV/Canvas](https://groups.google.com/g/bonsai-users/c/9M2UgJ_J_Qo/m/2igZWE8rBgAJ).

**Reply by frr665** [reply]
Hi Bruno, 

Thanks for the tip! I tried to build a full visualizer by replicating the workflow that you linked in your answer. I wanted to superpose my fly thorax position on each frame but I'm stuck with these 2 errors : 
1. Buffer operator doesn't accept input from CreatPoint operator (No method overload error)
![image](https://github.com/bonsai-rx/bonsai/assets/117995542/110d5306-fdd1-487d-ac70-d5b5593b1a01)
 
 

2. If I deactivate the line with the error I got the following error on AddImage operator :
`Assertion failed: src.channels() == dst.channels() in function cvCopy, copy,cpp(574)`


![image](https://github.com/bonsai-rx/bonsai/assets/117995542/4bef880e-b7ac-4a11-af1b-181aa94fd135)

I would greatly appreciate some suggestions to fix these issues !

Thanks a lot for your time, 
Hugo 

[test_sleap_tracking_incub.zip](https://github.com/bonsai-rx/bonsai/files/11677674/test_sleap_tracking_incub.zip)

**Reply by bruno-f-cruz** [reply]
Your color space does not match between the created canvas and the image (one of them is likely grayscale i.e. 1 channel). Is there a reason for not using the screen-grabbing strategy?

**Reply by frr665** [reply]
Hi I'm not using the screen capture strategy because I was curious to build a visualizer myself. 
I also want to record 4 flies tracking videos simultaneously for real time 24 hours of experiments, and I feel it would be a bit tedious to place all 4 visualizers next to each other at the beginning of each experiment for the screen capture. I prefer to record 4 custom visualizers in one video using a concat operator. 


Anyway now it's working, I can record a video of my fly thorax tracking ! 

![image](https://github.com/bonsai-rx/bonsai/assets/117995542/4dd84800-f5cb-4342-bab5-19d1b295c510)


Thanks a lot for your help,
Hugo 


[test_fly_visualizer.zip](https://github.com/bonsai-rx/bonsai/files/11679449/test_fly_visualizer.zip)

**Reply by bruno-f-cruz** [reply]
Just keep in mind that these visualizations greatly impact the online performance of the workflow. If can afford to build them offline, it is usually better!

**Reply by frr665** [reply]
Ok I will keep that in mind, thanks a lot !
